File: lessons/04_03/TypeScript(JavaScript)/1_base_classes.ts
class Animal {
  constructor(private health: number, private speed: number, private environment: string) {}

  getHealth(): number {
      return this.health;
  }

  setHealth(health: number): void {
      this.health = health;
  }

  getSpeed(): number {
      return this.speed;
  }

  setSpeed(speed: number): void {
      this.speed = speed;
  }

  animalInfo(): void {
      console.log(`Animal lives in ${this.environment}`);
  }
}

class Pigeon extends Animal {
  static titlePrint(): void {
      console.log("\tPIGEON INFO:");
  }
}

class Dog extends Animal {
  static titlePrint(): void {
      console.log("\tDOG INFO:");
  }
}

class Dolphin extends Animal {
  static titlePrint(): void {
      console.log("\tDOLPHIN INFO:");
  }
}

const pigeon = new Pigeon(100, 40, "Air");
const dog = new Dog(90, 25, "Earth");
const dolphin = new Dolphin(65, 70, "Water");

Pigeon.titlePrint();
console.log(`Pigeon health: ${pigeon.getHealth()}`);
console.log(`Pigeon speed: ${pigeon.getSpeed()}`);
pigeon.animalInfo();
console.log();

Dog.titlePrint();
console.log(`Dog health: ${dog.getHealth()}`);
console.log(`Dog speed: ${dog.getSpeed()}`);
dog.animalInfo();
console.log();

Dolphin.titlePrint();
console.log(`Dolphin health: ${dolphin.getHealth()}`);
console.log(`Dolphin speed: ${dolphin.getSpeed()}`);
dolphin.animalInfo();
console.log();

File: lessons/04_03/Python/1_base_classes.py
# В Python нет типизации как в C++, но есть аннотаиця типов, чтобы указывать где какой тип передается и возвращается
# Так же, нет protected как такого, есть договоренность, что _val - это переменная protected
# За то есть private, который обозначается __val
# Все переменные указываются в конструкторе(__init__)
# Статичсные переменные указываются вне конструктора

class Animal:
    def __init__(self, health: int, speed: float, environment: str) -> None:
        self.__health = health
        self.__speed = speed
        self.__environment = environment

    @property
    def health(self) -> int:
        return self.__health

    @health.setter
    def health(self, health: int) -> None:
        self.__health = health

    @property
    def speed(self) -> float:
        return self.__speed

    @speed.setter
    def set_speed(self, speed: float) -> None:
        self.__speed = speed

    def animal_info(self) -> None:
        print(f"Animal lives in {self.__environment}")


class Pigeon(Animal):
    @staticmethod
    def title_print() -> None:
        print("\tPIGEON INFO:")


class Dog(Animal):
    @staticmethod
    def title_print() -> None:
        print("\tDOG INFO:")


class Dolphin(Animal):
    @staticmethod
    def title_print() -> None:
        print("\tDOLPHIN INFO:")


pigeon = Pigeon(100, 40, "Air")
dog = Dog(90, 25, "Earth")
dolphin = Dolphin(65, 70, "Water")

Pigeon.title_print()
print(f"Pigeon health: {pigeon.health}")
print(f"Pigeon speed: {pigeon.speed}")
pigeon.animal_info()
print()

Dog.title_print()
print(f"Dog health: {dog.health}")
print(f"Dog speed: {dog.speed}")
dog.animal_info()
print()

Dolphin.title_print()
print(f"Dolphin health: {dolphin.health}")
print(f"Dolphin speed: {dolphin.speed}")
dolphin.animal_info()
print()
